pkg/env/env.go

```diff
@@ -160,70 +172,134 @@ func (e *testEnv) AfterEachTest(funcs ...types.TestEnvFunc) types.Environment {
+// AfterEachFeature. This can be invoked with a nillable wg which defines how the test feature is being run.
+// If the wg is set to nil that means the test is being run in sequence and there is no reason to mark a
+// WaitGroup as done
+func (e *testEnv) processTestFeature(wg *sync.WaitGroup, t *testing.T, featureName string, feature types.Feature) {
+	if wg != nil {
+		defer wg.Done()
+	}
+
+	var err error
 
 	// execute each feature
 	beforeFeatureActions := e.getBeforeFeatureActions()
 	afterFeatureActions := e.getAfterFeatureActions()
-	for i, feature := range testFeatures {
-		// execute beforeFeature actions
-		for _, action := range beforeFeatureActions {
-			if e.ctx, err = action.runWithFeature(e.ctx, e.cfg, deepCopyFeature(feature)); err != nil {
-				t.Fatalf("BeforeEachTest failure: %s", err)
-			}
-		}
 
-		// execute feature test
-		featName := feature.Name()
-		if featName == "" {
-			featName = fmt.Sprintf("Feature-%d", i+1)
+	for _, action := range beforeFeatureActions {
+		if e.ctx, err = action.runWithFeature(e.ctx, e.cfg, deepCopyFeature(feature)); err != nil {
+			t.Fatalf("BeforeEachTest failure: %s", err)
 		}
-		e.ctx = e.execFeature(e.ctx, t, featName, feature)
+	}
 
-		// execute beforeFeature actions
-		for _, action := range afterFeatureActions {
-			if e.ctx, err = action.runWithFeature(e.ctx, e.cfg, deepCopyFeature(feature)); err != nil {
-				t.Fatalf("BeforeEachTest failure: %s", err)
-			}
+	// execute feature test
+	e.ctx = e.execFeature(e.ctx, t, featureName, feature)
+
+	// execute beforeFeature actions
+	for _, action := range afterFeatureActions {
+		if e.ctx, err = action.runWithFeature(e.ctx, e.cfg, deepCopyFeature(feature)); err != nil {
+			t.Fatalf("BeforeEachTest failure: %s", err)
 		}
 	}
+}
 
-	// execute afterTest functions
+// processTests is a wrapper function that can be invoked by either Test or TestInParallel methods.
+// Depending on the configuration of if the parallel tests are enabled or not, this will change the
+// nature of how the test gets executed.
+//
+// In case if the parallel run of test features are enabled, this function will invoke the processTestFeature
+// as a go-routine to get them to run in parallel
+func (e *testEnv) processTests(t *testing.T, enableParallelRun bool, testFeatures ...types.Feature) {
+	e.panicOnMissingContext()
+	if len(testFeatures) == 0 {
+		t.Log("No test testFeatures provided, skipping test")
+		return
+	}
+	beforeTestActions := e.getBeforeTestActions()
 	afterTestActions := e.getAfterTestActions()
-	for _, action := range afterTestActions {
-		if e.ctx, err = action.runWithT(e.ctx, e.cfg, t); err != nil {
-			t.Fatalf("AfterEachTest failure: %s", err)
+
+	e.processTestActions(t, beforeTestActions)
+
+	if e.cfg.ParallelTestEnabled() && enableParallelRun {
+		log.V(4).Info("Running test features in parallel")
+	}
+
+	var wg sync.WaitGroup
+	for i, feature := range testFeatures {
```

Comment:
Suggestion: I would separate the execution paths (parallel, non-parallel) cleanly for clarity.  Also, this would make method `e.processTestFeature` truly generic without the need to pass a nil.

```go
if parallelEnabled {
    var wg sync.WaitGroup
    for i, feature := range testFeatures {
        wg.Add(1)
        go func(w *sync.Workgroup){
            defer w.Done()
            e.processTestFeature(t, featName, feature)
          }(&wg)
    }
    wg.Wait()
}else{
    for i, feature := range testFeatures {
         e.processTestFeature(t, featName, feature)
    }
}
```

Reply:
You have a valid point. The `porcessTestFeature` can/should indeed be agnostic of the `WaitGroup` and the sequence/parallel workflow. What you suggested here looks better. Let me update it accordingly.

Reply:
```golang
for i, feature := range testFeatures {
		featName := feature.Name()
		if featName == "" {
			featName = fmt.Sprintf("Feature-%d", i+1)
		}
		if runInParallel {
			wg.Add(1)
			go func(wg *sync.WaitGroup) {
				defer wg.Done()
				e.processTestFeature(t, featName, feature)
			}(&wg)
		} else {
			e.processTestFeature(t, featName, feature)
		}
	}
	if runInParallel {
		wg.Wait()
	}
```

turned it like above to avoid duplication of loop + the feature name generation bit.